sql/core/src/main/scala/org/apache/spark/sql/execution/dynamicpruning/PartitionPruning.scala

```diff
@@ -247,13 +246,15 @@ object PartitionPruning extends Rule[LogicalPlan] with PredicateHelper with Join
             // otherwise the pruning will not trigger
             var partScan = getPartitionTableScan(l, left)
             if (partScan.isDefined && canPruneLeft(joinType) &&
-                hasPartitionPruningFilter(right)) {
+                hasPartitionPruningFilter(right) &&
+                (canBroadcastBySize(right, conf) || hintToBroadcastRight(hint))) {
```

Comment:
to confirm: the major change of this PR is to also consider broadcast hint?

Reply:
if it is, can we make it clear in the PR title and description, and add a test?

Reply:
Actually we already have a test:
https://github.com/apache/spark/blob/56b18386ab71566480777660f52d4913aa319a2f/sql/core/src/test/scala/org/apache/spark/sql/DynamicPartitionPruningSuite.scala#L1117
Anyway, I added a new test.